src/lua/compat.lua

```diff
@@ -0,0 +1,307 @@
+        local max_key_len = 0
+        for _, key in pairs(options_order) do
+            max_key_len = (#key > max_key_len) and #key or max_key_len
+        end
+        for _, key in pairs(options_order) do
+            local val
+            local comment
+            if options[key].obsolete then
+                if action == ACT_NEW or action == ACT_OLD or action == ACT_CURRENT or
+                        (action == ACT_NIL and options[key].selected) then
+                    val = "'new'"
+                else
+                    val = "'default'"
+                end
+                comment = 'obsolete since ' .. options[key].obsolete
+            else
+                if action == ACT_CURRENT then
+                    if options[key].selected and options[key].current == NEW or options[key].default == 'new' then
+                        val = "'new'"
+                    else
+                        val = "'old'"
+                    end
+                elseif action == ACT_DEFAULT or (action == ACT_NIL and not options[key].selected) then
+                    val = "'default'"
+                elseif action == ACT_NEW or (action == ACT_NIL and options[key].current == NEW) then
+                    val = "'new'"
+                else
+                    val = "'old'"
+                end
+            end
+            -- Can't use '\t' due to gh-7681.
+            result = ('%s\n    %s%s = %s,'):format(result, key, string.rep(' ', max_key_len - #key), val)
+            if comment then
+                if (val ~= "'default'") then
+                    -- For alignment.
+                    result = result .. '    '
+                end
+                result = result .. ' -- ' .. comment
+            end
+        end
+        if not result:find('\n') then
+            return result .. '})'
+        else
+            -- Need double '\n' due to gh-3012.
+            return result .. '\n})\n\n'
+        end
+    end,
+    add_option = function(policy)
+        if type(policy) ~= 'table' then
+            error("usage: compat.add_option({name = '...', default = 'new'/'old'," ..
+                  " brief = '...', action = func, run_action = true/false})")
+        end
+        local name = policy.name
+        verify_option(name, policy)
+
+        local current = policy.default == 'new'
+        if not options[name] then
+            options[name] = {
+                current = current,
+                selected = false
+            }
+            table.insert(options_order, name)
+        elseif not options[name].selected then
+            options[name].current = current
+        end
+
+        local option = options[name]
+        option.brief    = policy.brief
+        option.default  = policy.default
+        option.obsolete = policy.obsolete
+        option.action   = policy.action
+
+        if not option.obsolete and policy.run_action then
+            option.action(options[name].current)
+        end
```

Comment:
I looked at the API description in the RFC and I don't see anything that would say that the action will not be called for an obsolete option. However it looks meaningful, so I would consider it as the API description problem.

Reply:
Added remark in `add_option()` description.